File: C/203_RemoveLinkedListElements.c
/*
203. Remove Linked List Elements

Remove all elements from a linked list of integers that have value val.

Example
Given: 1 --> 2 --> 6 --> 3 --> 4 --> 5 --> 6, val = 6
Return: 1 --> 2 --> 3 --> 4 --> 5

Credits:
Special thanks to @mithmatt for adding this problem and creating all test cases.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 //4 ms
struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode* cur = head;
 
    while (head && (head->val == val))
    {
        head = cur = cur->next;                
    }
    
    while(cur)
    {       
        // Remove
        if (cur->next && (cur->next->val == val))
        {
            cur->next = cur->next->next;
        }
        cur = cur->next;      
    }    
    return head;
}

//Sample
struct ListNode* removeElements2(struct ListNode* head, int val) {
    
    if(!head)
        return NULL;
    while(head->val==val)
    {
        head=head->next;
        if(!head)
            return NULL;
    }
    struct ListNode* p=head;
    struct ListNode* t=p->next;
    while(t)
    {
        if(t->val==val)
        {
            p->next=t->next;
        }
        else
            p=t;
        t=t->next;
    }
    return head;
}

File: Java/203_RemoveLinkedListElements.java
// 1 ms
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) 
            return head;
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cur = dummy;
        
        while (cur != null && cur.next != null) {
            if (cur.next.val == val)
                cur.next = cur.next.next;
            else
                cur = cur.next;
        }
        
        return dummy.next;
    }
}